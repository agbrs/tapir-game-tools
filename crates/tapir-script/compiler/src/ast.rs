use crate::tokens::{FileId, LexicalError, LexicalErrorKind, Span};

use serde::Serialize;

#[derive(Clone, Debug, Serialize)]
pub struct Statement<'input> {
    pub span: Span,
    pub kind: StatementKind<'input>,
}

#[derive(Clone, Debug, Serialize)]
pub enum StatementKind<'input> {
    VariableDeclaration {
        ident: &'input str,
        value: Box<Expression<'input>>,
    },
    Assignment {
        ident: &'input str,
        value: Box<Expression<'input>>,
    },
    Wait,
}

impl<'input> StatementKind<'input> {
    pub fn with_span(self, file_id: FileId, start: usize, end: usize) -> Statement<'input> {
        Statement {
            span: Span::new(file_id, start, end),
            kind: self,
        }
    }
}

#[derive(Clone, Debug, Serialize)]
pub struct Expression<'input> {
    pub span: Span,
    pub kind: ExpressionKind<'input>,
}

impl<'input> From<LexicalError> for Box<Expression<'input>> {
    fn from(value: LexicalError) -> Self {
        Box::new(Expression {
            span: value.span,
            kind: ExpressionKind::Error(value.kind),
        })
    }
}

#[derive(Clone, Debug, Serialize)]
pub enum ExpressionKind<'input> {
    Integer(i32),
    Variable(&'input str),
    BinaryOperation {
        lhs: Box<Expression<'input>>,
        operator: BinaryOperator,
        rhs: Box<Expression<'input>>,
    },
    Error(LexicalErrorKind),
}

impl<'input> ExpressionKind<'input> {
    pub fn with_span(self, file_id: FileId, start: usize, end: usize) -> Box<Expression<'input>> {
        Box::new(Expression {
            kind: self,
            span: Span::new(file_id, start, end),
        })
    }
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub enum BinaryOperator {
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    RealDiv,
    RealMod,
}
