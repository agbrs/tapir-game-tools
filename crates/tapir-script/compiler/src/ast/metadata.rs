use std::{
    any::{type_name, Any, TypeId},
    collections::HashMap,
    sync::Arc,
};

use serde::Serialize;

#[derive(Default, Clone, Debug)]
pub(crate) struct Metadata {
    map: HashMap<TypeId, (Arc<dyn Any>, &'static str)>,
}

impl Serialize for Metadata {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut type_names: Vec<_> = self.map.values().map(|(_, name)| name).collect();
        type_names.sort_unstable();
        type_names.serialize(serializer)
    }
}

impl Metadata {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn set<T: 'static>(&mut self, value: T) {
        let arced: Arc<dyn Any> = Arc::new(value);
        let type_id = arced.type_id();

        let type_name = type_name::<T>();

        self.map.insert(type_id, (arced, type_name));
    }

    pub fn get<T: 'static>(&self) -> Option<&T> {
        let type_id = TypeId::of::<T>();
        self.map
            .get(&type_id)
            .and_then(|(value, _)| value.downcast_ref())
    }
}
