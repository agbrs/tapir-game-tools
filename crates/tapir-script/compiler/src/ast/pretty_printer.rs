use std::fmt::{Display, Write};

use super::{Expression, ExpressionKind, Function, Statement, StatementKind};

pub(super) fn pretty_print(function: &Function, output: &mut dyn Write) -> std::fmt::Result {
    if function.name == "@toplevel" {
        if !function.meta.is_empty() {
            writeln!(output, "# @toplevel: {:?}\n", function.meta)?;
        }
        pretty_print_statements(&function.statements, output, Indent(0))?;
    } else {
        writeln!(output)?;
        if !function.meta.is_empty() {
            writeln!(output, "# {:?}", function.meta)?;
        }

        write!(
            output,
            "{}fn {}(",
            if function.modifiers.is_event_handler.is_some() {
                "event "
            } else {
                ""
            },
            function.name
        )?;

        for argument in &function.arguments {
            write!(output, "{},", argument.t.t)?;
        }

        write!(output, ")")?;

        if !function.return_types.types.is_empty() {
            write!(output, " -> (")?;
            for return_type in &function.return_types.types {
                write!(output, "{},", return_type.t)?;
            }
            write!(output, ")")?;
        }

        writeln!(output, " {{")?;

        pretty_print_statements(&function.statements, output, Indent(1))?;
        writeln!(output, "}}")?;
    }

    Ok(())
}

fn pretty_print_statements(
    statements: &[Statement],
    output: &mut dyn Write,
    indent: Indent,
) -> std::fmt::Result {
    for statement in statements {
        if statement.meta.is_empty() {
        } else {
            writeln!(output, "{indent}# {:?}", statement.meta)?;
        }

        write!(output, "{indent}")?;
        match &statement.kind {
            StatementKind::Error => write!(output, "ERROR;")?,
            StatementKind::VariableDeclaration { ident, value } => {
                write!(output, "var {ident} = ")?;
                pretty_print_expr(value, output, indent.increase())?;
                write!(output, ";")?;
            }
            StatementKind::Assignment { ident, value } => {
                write!(output, "{ident} = ")?;
                pretty_print_expr(value, output, indent.increase())?;
                write!(output, ";")?;
            }
            StatementKind::Wait => write!(output, "wait;")?,
            StatementKind::Block { block } => {
                writeln!(output, "{{")?;
                pretty_print_statements(block, output, indent.increase())?;
                write!(output, "{indent}}}")?;
            }
            StatementKind::Continue => write!(output, "continue;")?,
            StatementKind::Break => write!(output, "break;")?,
            StatementKind::Nop => write!(output, "# nop")?,
            StatementKind::If {
                condition,
                true_block,
                false_block,
            } => {
                write!(output, "if ")?;
                pretty_print_expr(condition, output, indent.increase())?;
                writeln!(output, " {{")?;
                pretty_print_statements(true_block, output, indent.increase())?;

                if !false_block.is_empty() {
                    writeln!(output, "{indent}}} else {{")?;
                    pretty_print_statements(false_block, output, indent.increase())?;
                }
                write!(output, "{indent}}}")?;
            }
            StatementKind::Loop { block } => {
                writeln!(output, "loop {{")?;
                pretty_print_statements(block, output, indent.increase())?;
                write!(output, "{indent}}}")?;
            }
            StatementKind::Call { name, arguments } => {
                write!(output, "{name}(")?;
                for argument in arguments {
                    pretty_print_expr(argument, output, indent.increase())?;
                    write!(output, ",")?;
                }
                write!(output, ");")?;
            }
            StatementKind::Spawn { name, arguments } => {
                write!(output, "spawn {name}(")?;
                for argument in arguments {
                    pretty_print_expr(argument, output, indent.increase())?;
                    write!(output, ",")?;
                }
                write!(output, ");")?;
            }
            StatementKind::Return { values } => {
                write!(output, "return (")?;
                for value in values {
                    pretty_print_expr(value, output, indent.increase())?;
                    write!(output, ",")?;
                }
                write!(output, ");")?;
            }
            StatementKind::Trigger { name, arguments } => {
                write!(output, "trigger {name}(")?;
                for arg in arguments {
                    pretty_print_expr(arg, output, indent.increase())?;
                    write!(output, ",")?;
                }
                write!(output, ");")?;
            }
        }

        writeln!(output)?;
    }

    Ok(())
}

fn pretty_print_expr(
    expression: &Expression,
    output: &mut dyn Write,
    mut indent: Indent,
) -> std::fmt::Result {
    if !expression.meta.is_empty() {
        indent = indent.increase();
        write!(output, "\n{indent}")?;
    }

    match &expression.kind {
        ExpressionKind::Integer(int) => write!(output, "{int}")?,
        ExpressionKind::Fix(num) => write!(output, "{num}")?,
        ExpressionKind::Bool(boool) => write!(output, "{boool}")?,
        ExpressionKind::Variable(variable) => write!(output, "{variable}")?,
        ExpressionKind::BinaryOperation { lhs, operator, rhs } => {
            write!(output, "(")?;
            pretty_print_expr(lhs, output, indent)?;
            write!(
                output,
                " {} ",
                match operator {
                    super::BinaryOperator::Add => "+",
                    super::BinaryOperator::Sub => "-",
                    super::BinaryOperator::Mul => "*",
                    super::BinaryOperator::Div => "/",
                    super::BinaryOperator::Mod => "%",
                    super::BinaryOperator::RealDiv => "//",
                    super::BinaryOperator::RealMod => "%%",
                    super::BinaryOperator::FixMul => "*",
                    super::BinaryOperator::FixDiv => "/",
                    super::BinaryOperator::EqEq => "==",
                    super::BinaryOperator::NeEq => "!=",
                    super::BinaryOperator::Gt => ">",
                    super::BinaryOperator::GtEq => ">=",
                    super::BinaryOperator::Lt => "<",
                    super::BinaryOperator::LtEq => "<=",
                    super::BinaryOperator::Then => "then",
                }
            )?;
            pretty_print_expr(rhs, output, indent)?;
            write!(output, ")")?;
        }
        ExpressionKind::Error => write!(output, "ERROR")?,
        ExpressionKind::Nop => write!(output, "NOP")?,
        ExpressionKind::Call { name, arguments } => {
            write!(output, "{name}(")?;
            for argument in arguments {
                pretty_print_expr(argument, output, indent.increase())?;
                write!(output, ",")?;
            }
            write!(output, ")")?;
        }
    }

    if !expression.meta.is_empty() {
        write!(output, " # {:?}\n{indent}", expression.meta)?;
    }

    Ok(())
}

#[derive(Clone, Copy)]
struct Indent(usize);
impl Display for Indent {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        for _ in 0..self.0 {
            write!(f, "    ")?
        }

        Ok(())
    }
}

impl Indent {
    pub fn increase(self) -> Self {
        Self(self.0 + 1)
    }
}
