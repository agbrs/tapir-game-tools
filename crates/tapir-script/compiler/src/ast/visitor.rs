use std::mem;

use crate::{tokens::Span, Message};

use super::{BinaryOperator, Expression, ExpressionKind, Statement, StatementKind, SymbolId};

pub type VResult<T> = Result<T, Message>;

pub trait Visitable<T: Default> {
    fn visit(&mut self, v: &mut impl Visitor<T>) -> VResult<T>;
}

pub trait Visitor<T>
where
    T: Default,
    Self: Sized,
{
    fn visit_statement_error<'input>(&mut self, error: Message) -> VResult<StatementKind<'input>> {
        Ok(StatementKind::Error(error))
    }

    fn visit_variable_declaration<'input>(
        &mut self,
        ident: &'input str,
        mut value: Box<Expression<'input>>,
        _span: Span,
    ) -> VResult<StatementKind<'input>> {
        value.visit(self)?;

        Ok(StatementKind::VariableDeclaration { ident, value })
    }

    fn visit_assignment<'input>(
        &mut self,
        ident: &'input str,
        mut value: Box<Expression<'input>>,
        _span: Span,
    ) -> VResult<StatementKind<'input>> {
        value.visit(self)?;

        Ok(StatementKind::Assignment { ident, value })
    }

    fn visit_wait<'input>(&mut self, _span: Span) -> VResult<StatementKind<'input>> {
        Ok(StatementKind::Wait)
    }

    fn visit_integer<'input>(&mut self, i: i32, _span: Span) -> VResult<ExpressionKind<'input>> {
        Ok(ExpressionKind::Integer(i))
    }

    fn visit_variable<'input>(
        &mut self,
        ident: &'input str,
        _span: Span,
    ) -> VResult<ExpressionKind<'input>> {
        Ok(ExpressionKind::Variable(ident))
    }

    fn visit_binop<'input>(
        &mut self,
        mut lhs: Box<Expression<'input>>,
        operator: BinaryOperator,
        mut rhs: Box<Expression<'input>>,
        _span: Span,
    ) -> VResult<ExpressionKind<'input>> {
        lhs.visit(self)?;
        rhs.visit(self)?;

        Ok(ExpressionKind::BinaryOperation { lhs, operator, rhs })
    }

    fn visit_expression_error<'input>(
        &mut self,
        error: Message,
    ) -> VResult<ExpressionKind<'input>> {
        Ok(ExpressionKind::Error(error))
    }

    fn visit_symbol_assign<'input>(
        &mut self,
        ident: SymbolId,
        mut value: Box<Expression<'input>>,
        _span: Span,
    ) -> VResult<StatementKind<'input>> {
        value.visit(self)?;
        Ok(StatementKind::SymbolAssign { ident, value })
    }

    fn visit_symbol_declare<'input>(
        &mut self,
        ident: SymbolId,
        mut value: Box<Expression<'input>>,
        _span: Span,
    ) -> VResult<StatementKind<'input>> {
        value.visit(self)?;
        Ok(StatementKind::SymbolDeclare { ident, value })
    }

    fn visit_symbol<'input>(
        &mut self,
        ident: SymbolId,
        _span: Span,
    ) -> VResult<ExpressionKind<'input>> {
        Ok(ExpressionKind::Symbol(ident))
    }
}

impl<'input, T> Visitable<T> for Statement<'input>
where
    T: Default,
{
    fn visit(&mut self, v: &mut impl Visitor<T>) -> VResult<T> {
        self.kind = match mem::take(&mut self.kind) {
            StatementKind::Error(message) => v.visit_statement_error(message)?,
            StatementKind::VariableDeclaration { ident, value } => {
                v.visit_variable_declaration(ident, value, self.span)?
            }
            StatementKind::Assignment { ident, value } => {
                v.visit_assignment(ident, value, self.span)?
            }
            StatementKind::Wait => v.visit_wait(self.span)?,
            StatementKind::Nop => StatementKind::Nop,

            StatementKind::SymbolAssign { ident, value } => {
                v.visit_symbol_assign(ident, value, self.span)?
            }
            StatementKind::SymbolDeclare { ident, value } => {
                v.visit_symbol_declare(ident, value, self.span)?
            }
        };

        Ok(T::default())
    }
}

impl<'input, T> Visitable<T> for Vec<Statement<'input>>
where
    T: Default,
{
    fn visit(&mut self, v: &mut impl Visitor<T>) -> VResult<T> {
        for statement in self {
            statement.visit(v)?;
        }

        Ok(T::default())
    }
}

impl<'input, T> Visitable<T> for Expression<'input>
where
    T: Default,
{
    fn visit(&mut self, v: &mut impl Visitor<T>) -> VResult<T> {
        self.kind = match mem::take(&mut self.kind) {
            ExpressionKind::Integer(i) => v.visit_integer(i, self.span)?,
            ExpressionKind::Variable(ident) => v.visit_variable(ident, self.span)?,
            ExpressionKind::BinaryOperation { lhs, operator, rhs } => {
                v.visit_binop(lhs, operator, rhs, self.span)?
            }
            ExpressionKind::Error(message) => v.visit_expression_error(message)?,
            ExpressionKind::Nop => ExpressionKind::Nop,

            ExpressionKind::Symbol(symbol_id) => v.visit_symbol(symbol_id, self.span)?,
        };

        Ok(T::default())
    }
}
