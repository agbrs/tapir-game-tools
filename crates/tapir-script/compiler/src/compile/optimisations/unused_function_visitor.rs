use petgraph::prelude::*;

use serde::Serialize;

use crate::ast::{ExpressionKind, Function, FunctionId, Statement, StatementKind};

use super::ConstantOptimisationResult;

#[derive(Debug, Clone, Copy, Serialize)]
pub struct UnusedFunction;

pub fn unused_function_visitor(functions: &mut [Function]) -> ConstantOptimisationResult {
    let mut call_graph = DiGraphMap::new();

    let mut roots = vec![FunctionId(0)];

    for function in functions.iter_mut() {
        if function.meta.has::<UnusedFunction>() {
            continue; // don't need to inspect this since we already know it is unused
        }

        let function_id: FunctionId = *function.meta.get().unwrap();

        if function.modifiers.is_event_handler.is_some() {
            roots.push(function_id);
        }

        call_graph.add_node(function_id);

        visit_block(function_id, &function.statements, &mut call_graph);

        fn visit_block(
            calling_function: FunctionId,
            block: &[Statement],
            call_graph: &mut DiGraphMap<FunctionId, ()>,
        ) {
            macro_rules! visit_expr {
                ($expr:expr) => {
                    for expr in $expr.all_inner() {
                        if matches!(expr.kind, ExpressionKind::Call { .. }) {
                            if let Some(called_id) = expr.meta.get() {
                                call_graph.add_edge(calling_function, *called_id, ());
                            }
                        }
                    }
                };
            }

            macro_rules! visit_exprs {
                ($exprs:expr) => {
                    for expr in $exprs {
                        visit_expr!(expr);
                    }
                };
            }

            for statement in block {
                match &statement.kind {
                    StatementKind::Error
                    | StatementKind::Wait
                    | StatementKind::Continue
                    | StatementKind::Break
                    | StatementKind::Nop => continue,

                    StatementKind::VariableDeclaration { value, .. }
                    | StatementKind::Assignment { value, .. } => {
                        visit_expr!(value);
                    }
                    StatementKind::Loop { block } | StatementKind::Block { block } => {
                        visit_block(calling_function, block, call_graph)
                    }
                    StatementKind::If {
                        condition,
                        true_block,
                        false_block,
                    } => {
                        visit_expr!(condition);
                        visit_block(calling_function, true_block, call_graph);
                        visit_block(calling_function, false_block, call_graph);
                    }
                    StatementKind::Spawn { arguments, .. }
                    | StatementKind::Call { arguments, .. } => {
                        if let Some(called_id) = statement.meta.get() {
                            call_graph.add_edge(calling_function, *called_id, ());
                        }

                        visit_exprs!(arguments);
                    }
                    StatementKind::Return { values: arguments }
                    | StatementKind::Trigger { arguments, .. } => {
                        visit_exprs!(arguments);
                    }
                }
            }
        }
    }

    let mut called_functions = Dfs::empty(&call_graph);

    for root in roots {
        called_functions.move_to(root);
        while called_functions.next(&call_graph).is_some() {}
    }

    let called_functions = called_functions.discovered;

    let mut did_something = ConstantOptimisationResult::DidNothing;
    for function in functions.iter_mut() {
        let function_id: FunctionId = *function.meta.get().unwrap();

        if !called_functions.contains(&function_id) && !function.meta.set(UnusedFunction) {
            // we've newly set a meta field
            did_something = ConstantOptimisationResult::DidSomething;
        }
    }

    did_something
}

#[cfg(test)]
mod test {
    use std::fs;

    use insta::{assert_snapshot, glob};

    use crate::{
        compile::{
            loop_visitor::visit_loop_check, symtab_visitor::SymTabVisitor,
            type_visitor::TypeVisitor,
        },
        grammar,
        lexer::Lexer,
        reporting::Diagnostics,
        tokens::FileId,
        CompileSettings,
    };

    use super::*;

    #[test]
    fn unused_function_snapshot_tests() {
        glob!("snapshot_tests", "unused_function/*.tapir", |path| {
            let input = fs::read_to_string(path).unwrap();

            let lexer = Lexer::new(&input, FileId::new(0));
            let parser = grammar::ScriptParser::new();
            let file_id = FileId::new(0);

            let mut diagnostics = Diagnostics::new(file_id, path.file_name().unwrap(), &input);

            let mut script = parser.parse(file_id, &mut diagnostics, lexer).unwrap();

            let compile_settings = CompileSettings {
                properties: Vec::new(),
                enable_optimisations: true,
            };

            let mut symtab_visitor =
                SymTabVisitor::new(&compile_settings, &mut script.functions, &mut diagnostics);
            let mut type_visitor = TypeVisitor::new(&compile_settings, &script.functions);

            for function in &mut script.functions {
                visit_loop_check(function, &mut diagnostics);
                symtab_visitor.visit_function(function, &mut diagnostics);
                type_visitor.visit_function(
                    function,
                    symtab_visitor.get_symtab(),
                    &mut diagnostics,
                );
            }

            unused_function_visitor(&mut script.functions);

            let pretty_printed = script.pretty_print();

            assert_snapshot!(pretty_printed);
        });
    }
}
