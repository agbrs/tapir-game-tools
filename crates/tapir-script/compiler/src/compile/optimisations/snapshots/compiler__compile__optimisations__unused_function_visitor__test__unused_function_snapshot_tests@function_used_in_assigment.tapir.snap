---
source: crates/tapir-script/compiler/src/compile/optimisations/unused_function_visitor.rs
expression: pretty_printed
input_file: crates/tapir-script/compiler/src/compile/optimisations/snapshot_tests/unused_function/function_used_in_assigment.tapir
---
# @toplevel: {"FunctionId(0)"}

# {"SymbolId(0)"}
var y = 7;
wait;
# {"SymbolId(1)"}
var x = 
        foo() # {"FunctionId(1)"}
        ;
wait;
# {"SymbolId(1)"}
x = 121;
wait;

# {"FunctionId(1)"}
fn foo() -> (int,) {
    wait;
    return (33,);
}
